tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +227,280 @@ static bool HasCanonicalDimensionNumbers(
+    }
+
+    // Convert the representation from string/vector<char> to vector<string>
+    // i.e ("abc") -> {"a", "b", "c"} S.T we can have no limit on number of
+    // loops Since ASCII Chars has only 256 number of characters Find elilipsis
+    // in loop and replace and find number of dim/rank ellipsis represent then
+    // add batch dims into loop_dim as loop_dim.push_back({"0",... "N"}) where N
+    // is rank of batch then we can add the remaining dimensions. For example
+    // given config "ab...cde" representing where batch rank is 3 then we will
+    // get loop_dim = f("ab...cde") = {"a","b","0","1","2","c","d","e"}
+    std::vector<std::string> all_loops = {lhs_loop, rhs_loop, output_loop};
+    std::vector<std::vector<std::string>> all_loop_dim;
+    std::vector<int> batch_rank_vec;
+    size_t max_operand_sz = 3;  // 2 Inputs + 1 output operand
+    for (auto loop : llvm::enumerate(all_loops)) {
+      std::vector<std::string> loop_dim;
+      size_t pre_elip = loop.value().find(ellipsis);
+      bool has_elip = true;
+      if (pre_elip == std::string::npos) {
+        pre_elip = loop.value().size();
+        has_elip = false;
+      }
+      // Adding dim up to ellipsis if it exist
+      for (int pre_ch_ind = 0; pre_ch_ind < pre_elip; pre_ch_ind++)
+        loop_dim.push_back(std::string(1, loop.value()[pre_ch_ind]));
+      if (!has_elip) {
+        all_loop_dim.push_back(loop_dim);
+        continue;
+      }
+      // Case where Ellipsis presence:
+      size_t postElip = loop.value().size() - (pre_elip + ellipsis.size());
+      size_t operand_rank = loop.index() < (max_operand_sz - 1)
+                                ? get_rank(args[loop.index()])
+                                : result_ty->getRank();
+      size_t batch_rank = operand_rank - (pre_elip + postElip);
+      // Add ("0",...,"N") where N is rank of batch into the loop
+      for (int batch_ind = 0; batch_ind < batch_rank; batch_ind++)
+        loop_dim.push_back(std::to_string(batch_ind));
+      // Add the dimension after ellipsis into the loop
+      for (int post_ch_ind = pre_elip + ellipsis.size();
+           post_ch_ind < loop.value().size(); post_ch_ind++)
+        loop_dim.push_back(std::string(1, loop.value()[post_ch_ind]));
+      batch_rank_vec.push_back(batch_rank);
+      all_loop_dim.push_back(loop_dim);
+    }
+    // Check that all ellipsis represent same rank of batch
+    if (batch_rank_vec.size() > 1) {
+      if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                      batch_rank_vec.begin())) {
+        return rewriter.notifyMatchFailure(
+            op, "Invalid Elipsis(ellipsis) within Einsum config!");
+      }
+    }
+
+    // Find all unique indices in the input and output
+    std::set<std::string> input_ind;
+    std::set<std::string> output_ind;
+    int num_of_input_operands = args.size();
+    std::vector<std::vector<std::string>> input_loops_dim_vec = {
+        all_loop_dim[0], all_loop_dim[1]};
```

Comment:
This is making vectors copies for the sole purpose of iterating in the loop, seems wasteful?

Reply:
done